middleware/common/include/common/server/handle/service.h

```diff
@@ -90,9 +90,34 @@ namespace casual
 
          auto execute_reply = execute::scope( [&]()
          {
+            auto normalize_code = []( auto& reply) -> decltype(auto)
+            {
+               using Result = decltype( reply.code.result);
+               if( reply.code.result != Result::ok)
+                  return reply;
+
+               // if transaction state is _not good_ we need to indicate this on the
+               // normal "reply code channel".
+               switch( reply.transaction.state)
```

Comment:
Not 100% sure about exact circumstances when this is executed. And what the "context" is. Looking at this "locally" it seems to be all cases where a service returns with reply.code.result == Result::ok.
Is there always a transaction state?
What if a service is called without a transaction? Not all services have transactions active when called. A service may start its own transaction(s), or may not use transactions at all.

Reply:
Well, the actual impact, or rather what we are trying to achieve, is if the resource interaction fails we need to propagate this to caller. If xa_start fails, the (user) service code is not even invoked. From the caller perspective, the service failed, there's no question about that, right? I don't see how one could argue otherwise... This is stuff that happens outside user control. Either before service invokation, or after.

Reply:
Sure, if e.g. xa_start fails during "service invocation" (before user code is called) it is correct to return TPESVCERR. Also if something fails at the end (xa_end...). TPESVCERR is the correct value to return for errors that occur during service invocation and service return. (There is also an error TPETRAN described for tpcall that is for tpcall without TPNOTRAN to a service that does not support transactions. This is also "kind of" an error during invocation...)
   
I was worried about the case of a service invoked with tpcall(..., TPNOTRAN) for a service that has for example "transaction: join". What is reply.transaction.state in that case? I assume there is no "transaction" in that case and it is not immediately obvious that the value is == "Enum::active". It MAY be initialized to that even without transaction... I investigated and it seems as if it is...  (at common/include/common/message/service.h line 151). So the code should work...

Reply:
The other case I was thinking about was transactions started during service invocation ("transaction: auto" and no transaction passed from caller). The reasonable behaviour in that case is to commit the current transaction if the service returns TPSUCCESS, rollback if the service returns TPFAIL and TPESVCERR if an error occurs during commit (or rollback?).

I assume this is handled by execute_transaction().

The current transaction on tpreturn might be different from the one started during invocation if the service has called tx_commit & tx_begin while executing. This is probably "legal" if the transaction was started on service invocation... This is an unlikely case as a service with "auto" probably is coded with the expectation that it is invoked with a transaction active, and then it is not allowed to commit or rollback. It can try :-)

I wonder what happens if the (auto) service has called tx_commit or tx_rollback before tpreturn. This is unlikely in, but I have not investigated.